File: agenkit-cpp/examples/evaluation/basic_metrics.cpp
/**
 * @file basic_metrics.cpp
 * @brief Basic metrics collection and aggregation example
 *
 * This example demonstrates how to use the evaluation framework to:
 * - Create SessionResult instances to track agent sessions
 * - Add metric measurements (quality, cost, duration)
 * - Collect multiple session results
 * - Compute aggregate statistics across sessions
 *
 * This is the foundation for monitoring agent performance over time,
 * tracking success rates, detecting issues, and measuring improvements.
 *
 * Compile: See examples/CMakeLists.txt
 * Run: ./basic_metrics
 */

#include "agenkit/evaluation/metrics.hpp"
#include <iostream>
#include <iomanip>
#include <random>
#include <string>
#include <thread>
#include <chrono>

using namespace agenkit::evaluation;

// Simulate running an agent session and collecting metrics
SessionResult simulate_agent_session(const std::string& session_id, const std::string& agent_name) {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_real_distribution<> quality_dist(0.7, 1.0);
    static std::uniform_int_distribution<> token_dist(100, 500);
    static std::uniform_real_distribution<> duration_dist(0.5, 2.5);
    static std::uniform_real_distribution<> success_dist(0.0, 1.0);

    SessionResult result(session_id, agent_name);

    // Simulate some processing time
    std::this_thread::sleep_for(std::chrono::milliseconds(10 + (rand() % 50)));

    // Add quality metrics
    double quality_score = quality_dist(gen);
    result.add_metric_measurement(create_quality_metric(
        "response_quality",
        quality_score * 10,
        10.0,
        {{"evaluator", "rule_based"}}
    ));

    // Add cost metrics
    int tokens_used = token_dist(gen);
    double cost_per_token = 0.00001;
    double total_cost = tokens_used * cost_per_token;
    result.add_metric_measurement(create_cost_metric(
        total_cost,
        "USD",
        {{"tokens", tokens_used}}
    ));

    // Add duration metrics
    double duration_seconds = duration_dist(gen);
    result.add_metric_measurement(create_duration_metric(duration_seconds));

    // Add custom success rate metric
    bool success = success_dist(gen) > 0.2; // 80% success rate
    double success_value = success ? 1.0 : 0.0;

    if (success) {
        result.set_status(SessionStatus::Completed);
    } else {
        result.set_status(SessionStatus::Failed);
        result.add_error("processing_error", "Failed to complete task", {{"reason", "timeout"}});
    }

    result.add_metric_measurement(MetricMeasurement(
        "success",
        success_value,
        MetricType::SuccessRate
    ));

    return result;
}

int main() {
    std::cout << "Basic Metrics Collection Example" << std::endl;
    std::cout << "=================================" << std::endl << std::endl;

    // Step 1: Create metrics collector
    std::cout << "Step 1: Creating Metrics Collector" << std::endl;
    std::cout << "-----------------------------------" << std::endl;
    MetricsCollector collector;
    std::cout << "✓ Metrics collector created" << std::endl << std::endl;

    // Step 2: Simulate multiple agent sessions
    std::cout << "Step 2: Simulating Agent Sessions" << std::endl;
    std::cout << "----------------------------------" << std::endl;
    int num_sessions = 20;
    std::cout << "Running " << num_sessions << " simulated agent sessions..." << std::endl << std::endl;

    for (int i = 0; i < num_sessions; ++i) {
        std::string session_id = "session-" + std::string(3 - std::to_string(i+1).length(), '0') + std::to_string(i+1);
        std::string agent_name = "example-agent";

        SessionResult result = simulate_agent_session(session_id, agent_name);
        collector.add_result(result);

        // Print progress
        std::string status = result.status() == SessionStatus::Completed ? "✓" : "✗";
        std::cout << "  " << status << " Session " << (i+1) << ": "
                  << session_status_to_string(result.status()) << std::endl;
    }
    std::cout << std::endl;

    // Step 3: Compute aggregate statistics
    std::cout << "Step 3: Computing Aggregate Statistics" << std::endl;
    std::cout << "---------------------------------------" << std::endl;
    auto stats = collector.get_statistics();

    std::cout << "Total Sessions: " << stats["session_count"] << std::endl;
    std::cout << "Completed: " << stats["completed_count"] << std::endl;
    std::cout << "Failed: " << stats["failed_count"] << std::endl;
    std::cout << std::fixed << std::setprecision(1);
    std::cout << "Success Rate: " << stats["success_rate"].get<double>() * 100 << "%" << std::endl;
    std::cout << std::fixed << std::setprecision(2);
    std::cout << "Average Duration: " << stats["avg_duration"] << "s" << std::endl;
    std::cout << "Total Errors: " << stats["total_errors"] << std::endl;
    std::cout << "Avg Errors/Session: " << stats["avg_errors_per_session"] << std::endl << std::endl;

    // Step 4: Analyze specific metrics
    std::cout << "Step 4: Analyzing Specific Metrics" << std::endl;
    std::cout << "-----------------------------------" << std::endl;

    // Quality metrics
    auto quality_stats = collector.get_metric_aggregates("response_quality");
    if (quality_stats["count"].get<int>() > 0) {
        std::cout << "\nQuality Metrics:" << std::endl;
        std::cout << "  Count: " << quality_stats["count"] << std::endl;
        std::cout << std::fixed << std::setprecision(3);
        std::cout << "  Mean: " << quality_stats["mean"] << std::endl;
        std::cout << "  Min: " << quality_stats["min"] << std::endl;
        std::cout << "  Max: " << quality_stats["max"] << std::endl;
    }

    // Cost metrics
    auto cost_stats = collector.get_metric_aggregates("total_cost");
    if (cost_stats["count"].get<int>() > 0) {
        std::cout << "\nCost Metrics:" << std::endl;
        std::cout << "  Count: " << cost_stats["count"] << std::endl;
        std::cout << std::fixed << std::setprecision(4);
        std::cout << "  Total Cost: $" << cost_stats["sum"] << std::endl;
        std::cout << "  Average Cost/Session: $" << cost_stats["mean"] << std::endl;
        std::cout << "  Min Cost: $" << cost_stats["min"] << std::endl;
        std::cout << "  Max Cost: $" << cost_stats["max"] << std::endl;
    }

    // Duration metrics
    auto duration_stats = collector.get_metric_aggregates("duration");
    if (duration_stats["count"].get<int>() > 0) {
        std::cout << "\nDuration Metrics:" << std::endl;
        std::cout << "  Count: " << duration_stats["count"] << std::endl;
        std::cout << std::fixed << std::setprecision(2);
        std::cout << "  Total Duration: " << duration_stats["sum"] << "s" << std::endl;
        std::cout << "  Average Duration: " << duration_stats["mean"] << "s" << std::endl;
        std::cout << "  Min Duration: " << duration_stats["min"] << "s" << std::endl;
        std::cout << "  Max Duration: " << duration_stats["max"] << "s" << std::endl;
    }

    // Success rate metrics
    auto success_stats = collector.get_metric_aggregates("success");
    if (success_stats["count"].get<int>() > 0) {
        std::cout << "\nSuccess Rate Metrics:" << std::endl;
        std::cout << "  Count: " << success_stats["count"] << std::endl;
        std::cout << std::fixed << std::setprecision(1);
        std::cout << "  Success Rate: " << success_stats["mean"].get<double>() * 100 << "%" << std::endl;
    }

    // Step 5: Summary and best practices
    std::cout << "\n" << std::string(70, '=') << std::endl;
    std::cout << "Summary: Basic Metrics Collection" << std::endl;
    std::cout << std::string(70, '=') << std::endl;

    std::cout << "\nKey Capabilities:" << std::endl;
    std::cout << "1. SessionResult: Track individual agent session metrics" << std::endl;
    std::cout << "2. MetricsCollector: Aggregate metrics across multiple sessions" << std::endl;
    std::cout << "3. Metric Types: Quality, cost, duration, success rate, custom" << std::endl;
    std::cout << "4. Statistics: Success rate, averages, min/max, error rates" << std::endl;

    std::cout << "\nMetric Types Available:" << std::endl;
    std::cout << "- SuccessRate: Binary success/failure tracking" << std::endl;
    std::cout << "- QualityScore: Normalized quality scores (0.0-1.0)" << std::endl;
    std::cout << "- Cost: Token costs and API expenses" << std::endl;
    std::cout << "- Duration: Execution time tracking" << std::endl;
    std::cout << "- ErrorRate: Error frequency analysis" << std::endl;
    std::cout << "- TaskCompletion: Task completion tracking" << std::endl;
    std::cout << "- Custom: Domain-specific metrics" << std::endl;

    std::cout << "\nThread Safety:" << std::endl;
    std::cout << "MetricsCollector is thread-safe and can be used concurrently" << std::endl;
    std::cout << "from multiple threads without additional synchronization." << std::endl;

    std::cout << "\nBest Practices:" << std::endl;
    std::cout << "1. Create one SessionResult per agent invocation" << std::endl;
    std::cout << "2. Add measurements as they occur (streaming metrics)" << std::endl;
    std::cout << "3. Set final status (completed/failed) when session ends" << std::endl;
    std::cout << "4. Use helper functions (create_quality_metric, create_cost_metric)" << std::endl;
    std::cout << "5. Collect across many sessions for statistical significance" << std::endl;
    std::cout << "6. Export to JSON for long-term storage and analysis" << std::endl;

    return 0;
}

File: agenkit-ts/examples/evaluation/basic-metrics.ts
/**
 * Basic Metrics Collection Example
 *
 * This example demonstrates how to use the evaluation framework to:
 * - Create SessionResult instances to track agent sessions
 * - Add metric measurements (quality, cost, duration)
 * - Collect multiple session results
 * - Compute aggregate statistics across sessions
 *
 * This is the foundation for monitoring agent performance over time,
 * tracking success rates, detecting issues, and measuring improvements.
 *
 * Run with: npx tsx examples/evaluation/basic-metrics.ts
 */

import {
  SessionResult,
  MetricsCollector,
  SessionStatus,
  MetricType,
  createMetricMeasurement,
  createErrorRecord,
} from '../../src/evaluation';

/**
 * Helper functions for creating common metrics.
 */

function createQualityMetric(
  name: string,
  score: number,
  maxScore: number,
  metadata?: Record<string, unknown>
) {
  return createMetricMeasurement(
    name,
    score / maxScore,
    MetricType.QualityScore,
    metadata
  );
}

function createCostMetric(
  cost: number,
  currency: string,
  metadata?: Record<string, unknown>
) {
  return createMetricMeasurement('total_cost', cost, MetricType.Cost, {
    currency,
    ...metadata,
  });
}

function createDurationMetric(
  durationSeconds: number,
  metadata?: Record<string, unknown>
) {
  return createMetricMeasurement(
    'duration',
    durationSeconds,
    MetricType.Duration,
    metadata
  );
}

/**
 * Simulate running an agent session and collecting metrics.
 */
async function simulateAgentSession(
  sessionId: string,
  agentName: string
): Promise<SessionResult> {
  const result = new SessionResult(sessionId, agentName);

  // Simulate some processing time
  await new Promise(resolve => setTimeout(resolve, 10 + Math.random() * 50));

  // Add quality metrics
  const qualityScore = 0.7 + Math.random() * 0.3; // 0.7-1.0
  result.addMetricMeasurement(
    createQualityMetric('response_quality', qualityScore * 10, 10.0, {
      evaluator: 'rule_based',
    })
  );

  // Add cost metrics
  const tokensUsed = 100 + Math.floor(Math.random() * 400);
  const costPerToken = 0.00001;
  const totalCost = tokensUsed * costPerToken;
  result.addMetricMeasurement(
    createCostMetric(totalCost, 'USD', {
      tokens: tokensUsed,
    })
  );

  // Add duration metrics
  const durationSeconds = 0.5 + Math.random() * 2.0; // 0.5-2.5 seconds
  result.addMetricMeasurement(createDurationMetric(durationSeconds, {}));

  // Add custom success rate metric
  const success = Math.random() > 0.2; // 80% success rate
  const successValue = success ? 1.0 : 0.0;

  if (success) {
    result.setStatus(SessionStatus.Completed);
  } else {
    result.setStatus(SessionStatus.Failed);
    result.addError(
      createErrorRecord('processing_error', 'Failed to complete task', {
        reason: 'timeout',
      })
    );
  }

  result.addMetricMeasurement(
    createMetricMeasurement('success', successValue, MetricType.SuccessRate, {})
  );

  return result;
}

async function main() {
  console.log('Basic Metrics Collection Example');
  console.log('=================================\n');

  // Step 1: Create metrics collector
  console.log('Step 1: Creating Metrics Collector');
  console.log('-----------------------------------');
  const collector = new MetricsCollector();
  console.log('✓ Metrics collector created\n');

  // Step 2: Simulate multiple agent sessions
  console.log('Step 2: Simulating Agent Sessions');
  console.log('----------------------------------');
  const numSessions = 20;
  console.log(`Running ${numSessions} simulated agent sessions...\n`);

  for (let i = 0; i < numSessions; i++) {
    const sessionId = `session-${String(i + 1).padStart(3, '0')}`;
    const agentName = 'example-agent';

    const result = await simulateAgentSession(sessionId, agentName);
    collector.addSession(result);

    // Print progress
    const status = result.status === SessionStatus.Completed ? '✓' : '✗';
    console.log(`  ${status} Session ${i + 1}: ${result.status}`);
  }
  console.log();

  // Step 3: Compute aggregate statistics
  console.log('Step 3: Computing Aggregate Statistics');
  console.log('---------------------------------------');

  const sessionCount = collector.getSessionCount();
  const completedCount = collector.getSessionsByStatus(SessionStatus.Completed).length;
  const failedCount = collector.getSessionsByStatus(SessionStatus.Failed).length;
  const successRate = collector.getSuccessRate();
  const avgDuration = collector.getAverageDuration();
  const totalErrors = collector.getTotalErrorCount();
  const avgErrorsPerSession = sessionCount > 0 ? totalErrors / sessionCount : 0;

  console.log(`Total Sessions: ${sessionCount}`);
  console.log(`Completed: ${completedCount}`);
  console.log(`Failed: ${failedCount}`);
  console.log(`Success Rate: ${(successRate * 100).toFixed(1)}%`);
  console.log(`Average Duration: ${avgDuration.toFixed(2)}s`);
  console.log(`Total Errors: ${totalErrors}`);
  console.log(`Avg Errors/Session: ${avgErrorsPerSession.toFixed(2)}\n`);

  // Step 4: Analyze specific metrics
  console.log('Step 4: Analyzing Specific Metrics');
  console.log('-----------------------------------');

  const aggregated = collector.getAggregatedMetrics();

  // Quality metrics
  const qualityStats = aggregated.get('response_quality');
  if (qualityStats && qualityStats.count > 0) {
    console.log('\nQuality Metrics:');
    console.log(`  Count: ${qualityStats.count}`);
    console.log(`  Mean: ${qualityStats.mean.toFixed(3)}`);
    console.log(`  Min: ${qualityStats.min.toFixed(3)}`);
    console.log(`  Max: ${qualityStats.max.toFixed(3)}`);
  }

  // Cost metrics
  const costStats = aggregated.get('total_cost');
  if (costStats && costStats.count > 0) {
    const totalCost = costStats.mean * costStats.count;
    console.log('\nCost Metrics:');
    console.log(`  Count: ${costStats.count}`);
    console.log(`  Total Cost: $${totalCost.toFixed(4)}`);
    console.log(`  Average Cost/Session: $${costStats.mean.toFixed(4)}`);
    console.log(`  Min Cost: $${costStats.min.toFixed(4)}`);
    console.log(`  Max Cost: $${costStats.max.toFixed(4)}`);
  }

  // Duration metrics
  const durationStats = aggregated.get('duration');
  if (durationStats && durationStats.count > 0) {
    const totalDuration = durationStats.mean * durationStats.count;
    console.log('\nDuration Metrics:');
    console.log(`  Count: ${durationStats.count}`);
    console.log(`  Total Duration: ${totalDuration.toFixed(2)}s`);
    console.log(`  Average Duration: ${durationStats.mean.toFixed(2)}s`);
    console.log(`  Min Duration: ${durationStats.min.toFixed(2)}s`);
    console.log(`  Max Duration: ${durationStats.max.toFixed(2)}s`);
  }

  // Success rate metrics
  const successStats = aggregated.get('success');
  if (successStats && successStats.count > 0) {
    console.log('\nSuccess Rate Metrics:');
    console.log(`  Count: ${successStats.count}`);
    console.log(`  Success Rate: ${(successStats.mean * 100).toFixed(1)}%`);
  }

  // Step 5: Examine individual session results
  console.log('\n\nStep 5: Examining Individual Sessions');
  console.log('--------------------------------------');
  const results = collector.getAllSessions();

  console.log('\nTop 3 Highest Quality Sessions:');
  console.log('-'.repeat(70));

  // Sort by quality
  const sessionsWithQuality = results
    .map(result => {
      const qualityMeasurement = result.getMeasurementsByName('response_quality')[0];
      return {
        result,
        quality: qualityMeasurement ? qualityMeasurement.value : 0,
      };
    })
    .sort((a, b) => b.quality - a.quality)
    .slice(0, 3);

  sessionsWithQuality.forEach((sw, i) => {
    const costMeasurement = sw.result.getMeasurementsByName('total_cost')[0];
    const durationMeasurement = sw.result.getMeasurementsByName('duration')[0];

    console.log(`${i + 1}. Session: ${sw.result.sessionId}`);
    console.log(`   Quality: ${sw.quality.toFixed(3)}`);
    if (costMeasurement) {
      console.log(`   Cost: $${costMeasurement.value.toFixed(4)}`);
    }
    if (durationMeasurement) {
      console.log(`   Duration: ${durationMeasurement.value.toFixed(2)}s`);
    }
    console.log(`   Status: ${sw.result.status}\n`);
  });

  // Step 6: Summary and best practices
  console.log('='.repeat(70));
  console.log('Summary: Basic Metrics Collection');
  console.log('='.repeat(70));

  console.log('\nKey Capabilities:');
  console.log('1. SessionResult: Track individual agent session metrics');
  console.log('2. MetricsCollector: Aggregate metrics across multiple sessions');
  console.log('3. Metric Types: Quality, cost, duration, success rate, custom');
  console.log('4. Statistics: Success rate, averages, min/max, error rates');

  console.log('\nMetric Types Available:');
  console.log('- MetricType.SuccessRate: Binary success/failure tracking');
  console.log('- MetricType.QualityScore: Normalized quality scores (0.0-1.0)');
  console.log('- MetricType.Cost: Token costs and API expenses');
  console.log('- MetricType.Duration: Execution time tracking');
  console.log('- MetricType.ErrorRate: Error frequency analysis');
  console.log('- MetricType.TaskCompletion: Task completion tracking');
  console.log('- MetricType.Custom: Domain-specific metrics');

  console.log('\nBest Practices:');
  console.log('1. Create one SessionResult per agent invocation');
  console.log('2. Add measurements as they occur (streaming metrics)');
  console.log('3. Set final status (completed/failed) when session ends');
  console.log('4. Use helper functions for common metric types');
  console.log('5. Collect across many sessions for statistical significance');
  console.log('6. Export to JSON for long-term storage and analysis');

  console.log('\nReal-World Applications:');
  console.log('- Monitor agent success rates over time');
  console.log('- Track API costs and token usage');
  console.log('- Identify slow or expensive sessions');
  console.log('- Detect quality degradation');
  console.log('- A/B test different agent configurations');
  console.log('- Generate performance reports and dashboards');
}

main().catch(console.error);
